clip/xclip/xclip.man
.\" $Id$
.\"
.\" xclip.man - xclip manpage
.\" Copyright (C) 2001 Kim Saunders
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH XCLIP 1 
.SH NAME
xclip \- command line interface to X selections (clipboard)
.SH SYNOPSIS
.B xclip
[OPTION] [FILE]...
.SH DESCRIPTION
Reads from standard in, or from one or more files, and makes the data available as an X selection for pasting into X applications. Prints current X selection to standard out.
.TP
\fB\-i\fR, \fB\-in\fR
read text into X selection from standard input or files (default)
.TP
\fB\-o\fR, \fB\-out\fR
prints the selection to standard out (generally for piping to a file or program)
.TP
\fB\-f\fR, \fB\-filter\fR
when xclip is invoked in the in mode with output level set to silent (the defaults), the filter option will cause xclip to print the text piped to standard in back to standard out unmodified
.TP
\fB\-l\fR, \fB\-loops\fR
number of X selection requests (pastes into X applications) to wait for before exiting, with a value of 0 (default) causing xclip to wait for an unlimited number of requests until another application (possibly another invokation xclip) takes ownership of the selection
.TP
\fB\-d\fR, \fB\-display\fR
X display to use (e.g. "localhost:0"), xclip defaults to the value in $\fBDISPLAY\fR if this option is omitted
.TP
\fB\-h\fR, \fB\-help\fR
show quick summary of options
.TP
\fB\-selection\fR
specify which X selection to use, options are "primary" to use XA_PRIMARY (default), "secondary" for XA_SECONDARY or "clipboard" for XA_CLIPBOARD
.TP
\fB\-version\fR
show version information
.TP
\fB\-silent\fR
forks into the background to wait for requests, no informational output, errors only (default)
.TP
\fB\-quiet\fR
show informational messages on the terminal and run in the foreground
.TP
\fB\-verbose\fR
provide a running commentary of what xclip is doing

.PP
xclip reads text from standard in or files and makes it available to other X applications for pasting as an X selection (traditionally with the middle mouse button). It reads from all files specified, or from standard in if no files are specified. xclip can also print the contents of a selection to standard out with the
.B
-o
option.

xclip was designed to allow tighter integration of X applications and command line programs. The default action is to silently wait in the background for X selection requests (pastes) until another X application places data in the clipboard, at which point xclip exits silently. You can use the \fB\-verbose\fR option to see if and when xclip actually recieves selection requests from other X applications.

Options can be abreviated as long as they remain unambiguous. For example, it is possible to use \fB\-d\fR or \fB\-disp\fR instead of \fB\-display\fR. However, \fB\-v\fR couldn't be used because it is ambiguous (it could be short for \fB\-verbose\fR or \fB\-version\fR), so it would be interpreted as a filename.

Note that only the first character of the selection specified with the \fB\-selection\fR option is important. This means that "p", "sec" and "clip" would have the same effect as using "primary", "secondary" or "clipboard" respectively.

.SH EXAMPLES
.PP
I hate man pages without examples!

.B
uptime | xclip
.PP
Put your uptime in the X selection. Then middle click in an X application to paste.

.B xclip -loops 10 -verbose /etc/motd
.PP
Exit after /etc/motd (message of the day) has been pasted 10 times. Show how many selection requests (pastes) have been processed.

.B xclip -o > helloworld.c
.PP
Put the contents of the selection into a file.

.SH ENVIRONMENT
.TP
.SM
\fBDISPLAY\fR
X display to use if none is specified with the
.B
-display
option.

.SH REPORTING BUGS
Please report any bugs, problems, queries, experiences, etc. directly to the author.

.SH AUTHOR
Kim Saunders <kims@debian.org>
.br


debian/changelog
clip (1.1.16-1) stable; urgency=low

  * new release, clip-xml, clip-glade2, few new functions, some optimize.

 -- Uri Khnykin <uri@itk.ru>  Thu, 18 Apr 2006 01:00:00 +0400

clip (1.1.15-1) stable; urgency=low

  * new release, some bugs fixed, speed opti,ize, few new functions.

 -- Uri Khnykin <uri@itk.ru>  Thu, 16 Dec 2005 01:00:00 +0500

clip (1.1.14-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, few new functions, CLIP-UI added.

 -- Uri Khnykin <uri@itk.ru>  Thu, 05 Mar 2005 12:00:00 +0500

clip (1.1.13-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, few new functions.

 -- Uri Khnykin <uri@itk.ru>  Thu,  17 Dec 2004 17:00:00 +0500

clip (1.1.12-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, few new functions.

 -- Uri Khnykin <uri@itk.ru>  Thu,  16 Oct 2004 00:01:00 +0500

clip (1.1.11-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, CODB & COBRA new features.

 -- Uri Khnykin <uri@itk.ru>  Thu,  26 Jul 2004 14:00:00 +0500

clip (1.1.10-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, CODB & COBRA new features.

 -- Uri Khnykin <uri@itk.ru>  Thu,  02 May 2004 14:00:00 +0500

clip (1.1.9-1) stable; urgency=low

  * new release, some bugs fixed, GTK2 new functions, CODB & COBRA new features.

 -- Uri Khnykin <uri@itk.ru>  Thu,  27 Feb 2004 01:00:00 +0500

clip (1.1.8-1) stable; urgency=low

  * new release, many new docs, GTK2 new functions, bugs fixed

 -- Uri Khnykin <uri@itk.ru>  Thu,  02 Feb 2004 21:00:00 +0500

clip (1.1.7-1) stable; urgency=low

  * new release, COBrA initial, GTK2 new functions, bugs fixed

 -- Uri Khnykin <uri@itk.ru>  Thu,  21 Nov 2003 16:00:00 +0500

clip (1.1.6-1) stable; urgency=low

  * new release, GTK2 new functions, bugs fixed

 -- Paul Lasarev <paul@itk.ru>  Thu,  15 Nov 2003 19:00:00 +0500

clip (1.1.5-1) stable; urgency=low

  * new release, GTK2 added

 -- Paul Lasarev <paul@itk.ru>  Thu,  21 Jul 2003 19:00:00 +0500

clip (1.1.0-1) stable; urgency=low

  * new release, CODB first announce

 -- Paul Lasarev <paul@itk.ru>  Thu,  21 Jul 2003 19:00:00 +0500

clip (1.0.7-1) stable; urgency=low

  * new release

 -- Paul Lasarev <paul@itk.ru>  Thu,  3 Jul 2003 12:33:03 +0500

clip (1.0.6-1) stable; urgency=low

  * new release

 -- Paul Lasarev <paul@itk.ru>  Fri, 23 May 2003 18:53:49 +0500

clip (1.0-5) stable; urgency=low

  * new release

 -- Paul Lasarev <paul@sun.itk.ru>  Thu, 12 Apr 2003 17:42:21 +0400

clip (1.0-1) stable; urgency=low

  * new release

 -- Paul Lasarev <paul@sun.itk.ru>  Thu, 20 Feb 2003 17:42:21 +0400

clip (0.99-9) prestable; urgency=low

  * subversion

 -- Paul Lasarev <paul@sun.itk.ru>  Fri, 24 Jan 2003 16:09:09 +0400

clip (0.99-8) prestable; urgency=low

  * subversion
  * about 200 changes !!!!

 -- Paul Lasarev <paul@itk.ru>  Mon,  6 Jan 2003 15:00:00 +0400

clip (0.99-7) prestable; urgency=low

  * subversion
  * about 360 changes !!!!

 -- Paul Lasarev <paul@itk.ru>  Mon,  1 Dec 2002 11:54:58 +0400

clip (0.99-6) prestable; urgency=low

  * subversion
  * added detecting of unresolved symbols

 -- Paul Lasarev <paul@itk.ru>  Mon, 14 Oct 2002 13:02:57 +0500

clip (0.99-5) prestable; urgency=low

  * subversion

 -- Paul Lasarev <paul@itk.ru>  Sun, 31 Sep 2002 18:23:14 +0400

clip (0.99-4) prestable; urgency=low

  * subversion

 -- Paul Lasarev <paul@itk.ru>  Sun, 30 July 2002 17:45:14 +0400

clip (0.99-3) unstable; urgency=low

  * subversion

 -- Paul Lasarev <paul@itk.ru>  Sun, 26 May 2002 15:35:14 +0500

clip (0.99-2) unstable; urgency=low

  * subversion

 -- Paul Lasarev <paul@itk.ru>  Mon, 29 Apr 2002 18:55:12 +0500

clip (0.99-1) unstable; urgency=low

  * repackaging

 -- Paul Lasarev <paul@itk.ru>  Tue, 19 Mar 2002 15:05:44 +0400

clip (0.97-1) unstable; urgency=low

  * initial chinese (Big5) support

 -- Paul Lasarev <paul@itk.ru>  Wed, 16 Jan 2002 19:57:34 +0400

clip (0.96-3) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Wed,  9 Jan 2002 18:29:07 +0400

clip (0.96-1) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Wed, 26 Dec 2001 19:01:03 +0400

clip (0.95-1) unstable; urgency=low

  * new version
  * mouse support

 -- Paul Lasarev <paul@itk.ru>  Wed, 28 Nov 2001 16:42:35 +0400

clip (0.94-4) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Fri, 23 Nov 2001 10:34:55 +0400

clip (0.94-1) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Wed, 14 Nov 2001 10:23:24 +0400

clip (0.93-2) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Thu,  8 Nov 2001 18:19:13 +0400

clip (0.93-1) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Tue, 23 Oct 2001 16:36:01 +0500

clip (0.92-7) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Mon, 22 Oct 2001 13:05:09 +0500

clip (0.92-6) unstable; urgency=low

  * new version

 -- Paul Lasarev <paul@itk.ru>  Fri, 12 Oct 2001 11:21:08 +0500

clip (0.92-5) unstable; urgency=low

  * bugfixes

 -- Paul Lasarev <paul@itk.ru>  Mon, 24 Sep 2001 11:03:39 +0500

clip (0.92-4) unstable; urgency=low

  * pre-stable release

 -- Paul Lasarev <paul@itk.ru>  Wed, 19 Sep 2001 18:01:11 +0500

clip (0.92-1) unstable; urgency=low

  * split into runtime and devel packages

 -- Paul Lasarev <paul@itk.ru>  Tue, 28 Aug 2001 11:27:00 +0500

clip (0.91-1) unstable; urgency=low

  * standard shared libs in release

 -- Paul Lasarev <paul@itk.ru>  Tue, 28 Aug 2001 11:27:00 +0500

clip (0.9-1) unstable; urgency=low

  * new release

 -- Paul Lasarev <paul@itk.ru>  Mon, 27 Aug 2001 10:44:12 +0500

clip (0.8-1) unstable; urgency=low

  * add cliplibs

 -- Paul Lasarev <paul@itk.ru>  Sat,  9 Jun 2001 15:22:39 +0500

clip (0.7-1) unstable; urgency=low

  * updates

 -- Paul Lasarev <paul@itk.ru>  Tue, 29 May 2001 14:49:33 +0500

clip (0.6-1) unstable; urgency=low


clip/xclip/xclip.c
/*
 *  $Id$
 * 
 *  xclip.c - command line interface to X server selections 
 *  Copyright (C) 2001 Kim Saunders
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <X11/Xmu/Atoms.h>
#include "xcdef.h"
#include "xcprint.h"
#include "xclib.h"

/* command line option table for XrmParseCommand() */
XrmOptionDescRec opt_tab[11];

/* Options that get set on the command line */
int             sloop = 0;			/* number of loops */
char           *sdisp = NULL;			/* X display to connect to */
Atom            sseln = XA_PRIMARY;		/* X selection to work with */

/* Flags for command line options */
static int      fverb = OSILENT;		/* output level */
static int      fdiri = T;			/* direction is in */
static int      ffilt = F;			/* filter mode */

Display	       *dpy;				/* connection to X11 display */
XrmDatabase     opt_db = NULL;			/* database for options */

char          **fil_names;			/* names of files to read */
int             fil_number = 0;                 /* number of files to read */
int		fil_current = 0;
FILE*           fil_handle = NULL;

/* variables to hold Xrm database record and type */
XrmValue        rec_val;
char           *rec_typ;

int		tempi = 0;

/* Use XrmParseCommand to parse command line options to option variable */
static void doOptMain (int argc, char *argv[])
{
	/* Initialise resource manager and parse options into database */
	XrmInitialize();
	XrmParseCommand(
		&opt_db,
		opt_tab,
		sizeof(opt_tab) / sizeof(opt_tab[0]),
		XC_NAME,
		&argc,
		argv
	);
  
	/* set output level */
	if (
		XrmGetResource(
			opt_db,
			"xclip.olevel",
			"Xclip.Olevel",
			&rec_typ,
			&rec_val
		)
	)
	{
		/* set verbose flag according to option */
		if (strcmp(rec_val.addr, "S") == 0)
			fverb = OSILENT;
		if (strcmp(rec_val.addr, "Q") == 0)
			fverb = OQUIET;
		if (strcmp(rec_val.addr, "V") == 0)
			fverb = OVERBOSE;
	}
  
	/* set direction flag (in or out) */
	if (
		XrmGetResource(
			opt_db,
			"xclip.direction",
			"Xclip.Direction",
			&rec_typ,
			&rec_val
		)
	)
	{
		if (strcmp(rec_val.addr, "I") == 0)
			fdiri = T;
		if (strcmp(rec_val.addr, "O") == 0)
			fdiri = F;
	}
  
	/* set filter mode */
	if (
		XrmGetResource(
			opt_db,
			"xclip.filter",
			"Xclip.Filter",
			&rec_typ,
			&rec_val
		)
	)
	{
		/* filter mode only allowed in silent mode */
		if (fverb == OSILENT)
			ffilt = T;
	}
  
	/* check for -help and -version */
	if (
		XrmGetResource(
			opt_db,
			"xclip.print",
			"Xclip.Print",
			&rec_typ,
			&rec_val
		)
	)
	{
		if (strcmp(rec_val.addr, "H") == 0)
			prhelp(argv[0]);
		if (strcmp(rec_val.addr, "V") == 0)
			prversion();
	}
  
	/* check for -display */
	if (
		XrmGetResource(
			opt_db,
			"xclip.display",
			"Xclip.Display",
			&rec_typ,
			&rec_val
		)
	)
	{
		sdisp = rec_val.addr;
		if (fverb == OVERBOSE)	/* print in verbose mode only */
			fprintf(stderr, "Display: %s\n", sdisp);
	}
  
	/* check for -loops */
	if (
		XrmGetResource(
			opt_db,
			"xclip.loops",
			"Xclip.Loops",
			&rec_typ,
			&rec_val
		)
	)
	{
		sloop = atoi(rec_val.addr);
		if (fverb == OVERBOSE)	/* print in verbose mode only */
			fprintf(stderr, "Loops: %i\n", sloop);
	}

	/* Read remaining options (filenames) */
	while ( (fil_number + 1) < argc )
	{
		if (fil_number > 0)
		{
			fil_names = xcrealloc(
				fil_names,
				(fil_number + 1) * sizeof(char*)
			);
		} else
		{
			fil_names = xcmalloc(sizeof(char*));
		}
		fil_names[fil_number] = argv[fil_number + 1];
		fil_number++;
	}
}

/* process selection command line option */
static void doOptSel (void)
{
	/* set selection to work with */
	if (
		XrmGetResource(
			opt_db,
			"xclip.selection",
			"Xclip.Selection",
			&rec_typ,
			&rec_val
		)
	)
	{
		switch (tolower(rec_val.addr[0]))
		{
			case 'p':
				sseln = XA_PRIMARY;
				break;
			case 's':
				sseln = XA_SECONDARY;
				break;
			case 'c':
				sseln = XA_CLIPBOARD(dpy);
				break;
		}
    
		if (fverb == OVERBOSE)
		{
			fprintf(stderr, "Using selection: ");
	   
			if (sseln == XA_PRIMARY)
				fprintf(stderr, "XA_PRIMARY");
			if (sseln == XA_SECONDARY)
				fprintf(stderr, "XA_SECONDARY");
			if (sseln == XA_CLIPBOARD(dpy))
				fprintf(stderr, "XA_CLIPBOARD");

			fprintf(stderr, "\n");
		}
	}
}

int main (int argc, char *argv[])
{
	/* Declare variables */
	unsigned char *sel_buf;		/* buffer for selection data */
	unsigned long sel_len = 0;	/* length of sel_buf */
	unsigned long sel_all = 0;	/* allocated size of sel_buf */
	Window win;			/* Window */
	XEvent evt;			/* X Event Structures */
	int dloop = 0;			/* done loops counter */

	pid_t pid;			/* child pid if forking */

	/* set up option table. I can't figure out a better way than this to
	 * do it while sticking to pure ANSI C. The option and specifier
	 * members have a type of volatile char *, so they need to be allocated
	 * by strdup or malloc, you can't set them to a string constant at
	 * declare time, this is note pure ANSI C apparently, although it does
	 * work with gcc
	 */
	
	/* loop option entry */
	opt_tab[0].option 	=	xcstrdup("-loops");
	opt_tab[0].specifier	=	xcstrdup(".loops");
	opt_tab[0].argKind	=	XrmoptionSepArg;
	opt_tab[0].value	=	(XPointer) NULL;

	/* display option entry */
	opt_tab[1].option	=	xcstrdup("-display");
	opt_tab[1].specifier	=	xcstrdup(".display");
	opt_tab[1].argKind	=	XrmoptionSepArg;
	opt_tab[1].value	=	(XPointer) NULL;
				
	/* selection option entry */
	opt_tab[2].option 	=	xcstrdup("-selection");
	opt_tab[2].specifier	=	xcstrdup(".selection");
	opt_tab[2].argKind	=	XrmoptionSepArg;
	opt_tab[2].value	=	(XPointer) NULL;
		
	/* filter option entry */
	opt_tab[3].option	=	xcstrdup("-filter");
	opt_tab[3].specifier	=	xcstrdup(".filter");
	opt_tab[3].argKind	=	XrmoptionNoArg;	
	opt_tab[3].value	=	(XPointer) xcstrdup(ST);
		
	/* in option entry */
	opt_tab[4].option	=	xcstrdup("-in");
	opt_tab[4].specifier	=	xcstrdup(".direction");
	opt_tab[4].argKind	=	XrmoptionNoArg;
	opt_tab[4].value	=	(XPointer) xcstrdup("I");
		
	/* out option entry */
	opt_tab[5].option	=	xcstrdup("-out");
	opt_tab[5].specifier	=	xcstrdup(".direction");
	opt_tab[5].argKind	=	XrmoptionNoArg;
	opt_tab[5].value	=	(XPointer) xcstrdup("O");
		
	/* version option entry */
	opt_tab[6].option	=	xcstrdup("-version");
	opt_tab[6].specifier	=	xcstrdup(".print");
	opt_tab[6].argKind	=	XrmoptionNoArg;
	opt_tab[6].value	=	(XPointer) xcstrdup("V");
		
	/* help option entry */
	opt_tab[7].option	=	xcstrdup("-help");
	opt_tab[7].specifier	=	xcstrdup(".print");
	opt_tab[7].argKind	=	XrmoptionNoArg;
	opt_tab[7].value	=	(XPointer) xcstrdup("H");
		
	/* silent option entry */
	opt_tab[8].option	=	xcstrdup("-silent");
	opt_tab[8].specifier	=	xcstrdup(".olevel");
	opt_tab[8].argKind	=	XrmoptionNoArg;
	opt_tab[8].value	=	(XPointer) xcstrdup("S");
		
	/* quiet option entry */
	opt_tab[9].option	=	xcstrdup("-quiet");
	opt_tab[9].specifier	=	xcstrdup(".olevel");
	opt_tab[9].argKind	=	XrmoptionNoArg;
	opt_tab[9].value	=	(XPointer) xcstrdup("Q");
		
	/* verbose option entry */
	opt_tab[10].option	=	xcstrdup("-verbose");
	opt_tab[10].specifier	=	xcstrdup(".olevel");
	opt_tab[10].argKind	=	XrmoptionNoArg;
	opt_tab[10].value	=	(XPointer) xcstrdup("V");

	/* parse command line options */
	doOptMain(argc, argv);
   
	/* Connect to the X server. */
	if ( (dpy = XOpenDisplay(sdisp)) )
	{
		/* successful */
		if (fverb == OVERBOSE)
			fprintf(stderr, "Connected to X server.\n");
	} else
	{
		/* couldn't connect to X server. Print error and exit */
		errxdisplay(sdisp);
	}

	/* parse selection command line option */
	doOptSel();
  
	/* Create a window to trap events */
	win = XCreateSimpleWindow(
		dpy,
		DefaultRootWindow(dpy),
		0,
		0,
		1,
		1,
		0,
		0,
		0
	);

	/* get events about property changes */
	XSelectInput(dpy, win, PropertyChangeMask);
  
	if (fdiri)
	{
		/* in mode */
		sel_all = 16;		/* Reasonable ballpark figure */
		sel_buf = xcmalloc(sel_all * sizeof(char));

		/* Put chars into inc from stdin or files until we hit EOF */
		do {
			if (fil_number == 0)
			{
				/* read from stdin if no files specified */
				fil_handle = stdin;
			} else
			{
				if (
					(fil_handle = fopen(
						fil_names[fil_current],
						"r"
					)) == NULL
				)
				{
					errperror(
						3,
						argv[0],
						": ",
						fil_names[fil_current]
					);
					exit(EXIT_FAILURE);
				} else
				{
					/* file opened successfully. Print
					 * message (verbose mode only).
					 */
					if (fverb == OVERBOSE)
						fprintf(
							stderr,
							"Reading %s...\n",
							fil_names[fil_current]
						);
				}
			}

			fil_current++;
			while (!feof(fil_handle))
			{
				/* If sel_buf is full (used elems =
				 * allocated elems)
				 */
				if (sel_len == sel_all)
				{
					/* double the number of
					 * allocated elements
					 */
					sel_all *= 2;
					sel_buf = (unsigned char *)xcrealloc(
						sel_buf,
						sel_all * sizeof(char)
					);
				}
				sel_len += fread(
					sel_buf + sel_len,
					sizeof(char),
					sel_all - sel_len,
					fil_handle
				);
			}
		} while (fil_current < fil_number);

		/* if there are no files being read from (i.e., input
		 * is from stdin not files, and we are in filter mode,
		 * spit all the input back out to stdout
		 */
		if ((fil_number == 0) && ffilt)
			fwrite(sel_buf, sizeof(char), sel_len, stdout); 
    
		/* take control of the selection so that we receive
		 * SelectionRequest events from other windows
		 */
		XSetSelectionOwner(dpy, sseln, win, CurrentTime);

		/* fork into the background, exit parent process if we
		 * are in silent mode
		 */
		if (fverb == OSILENT)
		{
			pid = fork();
			/* exit the parent process; */
			if (pid)
				exit(EXIT_SUCCESS);
		}

		/* print a message saying what we're waiting for */
		if (fverb > OSILENT)
		{
			if (sloop == 1)
				fprintf(
					stderr,
					"Waiting for one selection request.\n"
				);
				
			if (sloop  < 1)
				fprintf(
					stderr,
					"Waiting for selection requests, Control-C to quit\n"
				);
				
			if (sloop  > 1)
				fprintf(
					stderr,
					"Waiting for %i selection requests, Control-C to quit\n",
					sloop
				);
		}
  
		/* loop and wait for the expected number of
		 * SelectionRequest events
		 */
		while (dloop < sloop || sloop < 1)
		{
			/* print messages about what we're waiting for
			 * if not in silent mode
			 */
			if (fverb > OSILENT)
			{
				if (sloop  > 1)
					fprintf(
						stderr,
						"  Waiting for selection request %i of %i.\n",
						dloop + 1,
						sloop
					);
					
				if (sloop == 1)
					fprintf(
						stderr,
						"  Waiting for a selection request.\n"
					);
					
				if (sloop  < 1)
					fprintf(
						stderr,
						"  Waiting for selection request number %i\n",
						dloop + 1
					);
			}
     
			/* wait for a SelectionRequest event */
			while (1)
			{
				static unsigned int clear = 0;
				static unsigned int context = XCLIB_XCIN_NONE;
				static unsigned long sel_pos = 0;
				static Window cwin;
				static Atom pty;
				int finished;

				XNextEvent(dpy, &evt);

				finished = xcin(
					dpy,
					&cwin,
					evt,
					&pty,
					sel_buf,
					sel_len,
					&sel_pos,
					&context
				);

				if (evt.type == SelectionClear)
					clear = 1;

				if ( (context == XCLIB_XCIN_NONE) && clear)
					exit(EXIT_SUCCESS);

				if (finished)
					break;
			}

			dloop++;	/* increment loop counter */
		}
	} else
	{
		unsigned int context = XCLIB_XCOUT_NONE;

		while (1)
		{
			/* only get an event if xcout() is doing something */
			if (context != XCLIB_XCOUT_NONE)
				XNextEvent(dpy, &evt);

			/* fetch the selection, or part of it */
			xcout(
				dpy,
				win,
				evt,
				sseln,
				&sel_buf,
				&sel_len,
				&context
			);

			/* only continue if xcout() is doing something */
			if (context == XCLIB_XCOUT_NONE)
				break;
		}
		
		if (sel_len)
		{
			/* only print the buffer out, and free it, if it's not
			 * empty
			 */
			fwrite(sel_buf, sizeof(char), sel_len, stdout);
			free(sel_buf);
		}
	}

	/* Disconnect from the X server */
	XCloseDisplay(dpy);

	/* exit */
	return(EXIT_SUCCESS);
}
